File: Server/Landscape.cs
/*
    This file is part of Armored Deathmatch by Hans Milling.

    Armored Deathmatch is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Armored Deathmatch is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Armored Deathmatch.  If not, see <http://www.gnu.org/licenses/>.
	
*/

using System;
using System.Collections.Generic;
using System.Text;
using System.Globalization;

namespace Server
{
  /// <summary>
  /// Simple landscape generator. Ported from Turbo Pascal code I wrote a long time ago for a similar DOS based Scorched Earth like 3D game.
  /// </summary>
 
  class Landscape
  {
    StringBuilder json = new StringBuilder();
    public double scale = 20000;
    public double[,] data;
    public int landscapesize;
    public static Landscape landscape;

    public Landscape(int size)
    {
      landscape = this;
      landscapesize = size - 1;
      Random random = new Random();
      //size of grid to generate, note this must be a
      //value 2^n+1
      double h = 2500.0; //the range (-h -> +h) for the average offset
      double SEED = 500.0; //an initial seed value for the corners of the data
      data = new double[size, size];
      //seed the data

      // Fill it with 0
      for (int y = 0; y < size; y++)
      {
        for (int x = 0; x < size; x++)
        {
          data[x, y] = 0;
        }
      }

      data[0, 0] = random.NextDouble() * (2 * SEED) - SEED;
      data[0, size - 1] = random.NextDouble() * (2 * SEED) - SEED;
      data[size - 1, 0] = random.NextDouble() * (2 * SEED) - SEED;
      data[size - 1, size - 1] = random.NextDouble() * (2 * SEED) - SEED;

      //side length is distance of a single square side
      //or distance of diagonal in diamond
      for (int sideLength = size - 1;
        //side length must be >= 2 so we always have
        //a new value (if its 1 we overwrite existing values
        //on the last iteration)
          sideLength >= 2;
        //each iteration we are looking at smaller squares
        //diamonds, and we decrease the variation of the offset
          sideLength /= 2, h /= 2.0)
      {
        //half the length of the side of a square
        //or distance from diamond center to one corner
        //(just to make calcs below a little clearer)
        int halfSide = sideLength / 2;

        //generate the new square values
        for (int x = 0; x < size - 1; x += sideLength)
        {
          for (int y = 0; y < size - 1; y += sideLength)
          {
            //x, y is upper left corner of square
            //calculate average of existing corners
            double avg = data[x, y] + //top left
            data[x + sideLength, y] + //top right
            data[x, y + sideLength] + //lower left
            data[x + sideLength, y + sideLength]; //lower right
            avg /= 4.0;

            //center is average plus random offset
            data[x + halfSide, y + halfSide] =
              //We calculate random value in range of 2h
              //and then subtract h so the end value is
              //in the range (-h, +h)
          avg + (random.NextDouble() * 2 * h) - h;
          }
        }

        //generate the diamond values
        //since the diamonds are staggered we only move x
        //by half side
        //NOTE: if the data shouldn't wrap then x < DATA_SIZE
        //to generate the far edge values
        for (int x = 0; x < size - 1; x += halfSide)
        {
          //and y is x offset by half a side, but moved by
          //the full side length
          //NOTE: if the data shouldn't wrap then y < DATA_SIZE
          //to generate the far edge values
          for (int y = (x + halfSide) % sideLength; y < size - 1; y += sideLength)
          {
            //x, y is center of diamond
            //note we must use mod  and add DATA_SIZE for subtraction 
            //so that we can wrap around the array to find the corners
            double avg =
              data[(x - halfSide + size) % size, y] + //left of center
              data[(x + halfSide) % size, y] + //right of center
              data[x, (y + halfSide) % size] + //below center
              data[x, (y - halfSide + size) % size]; //above center
            avg /= 4.0;

            //new value = average plus random offset
            //We calculate random value in range of 2h
            //and then subtract h so the end value is
            //in the range (-h, +h)
            avg = avg + (random.NextDouble() * 2 * h) - h;
            //update value for center of diamond
            data[x, y] = avg;

            //wrap values on the edges, remove
            //this and adjust loop condition above
            //for non-wrapping values.
            if (x == 0) data[size - 1, y] = avg;
            if (y == 0) data[x, size - 1] = avg;
          }
        }
      }
      json.Append("[");
      for (int x = 0; x < size; x++)
      {
        if (x != 0)
          json.Append(",");
        json.Append("[");
        for (int y = 0; y < size; y++)
        {
          //data[x, y] = Math.Sin((double)y/5.0)*2000.0;
          if (y != 0)
            json.Append(",");
          json.Append(data[x, y].ToString(CultureInfo.InvariantCulture));
        }
        json.Append("]");
      }
      json.Append("]");
    }

    public string toJSON()
    {
      return json.ToString();
    }

    private double getHeight(double px, double pz, int x1, int z1, int x2, int z2, int x3, int z3)
    {
      if (x1 > 256 || z1 > 256 || x2 > 256 || z2 > 256 || x3 > 256 || z3 > 256 || x1<0 ||x2<0||x3<0||z1<0||z2<0||z3<0)
        return 0;
      double h1 = data[x1, z1];
      double h2 = data[x2, z2];
      double h3 = data[x3, z3];
      double diffx = h1 - h2;
      double diffz = h3 - h2;
      double h = px * diffx + pz * diffz;
      return h2 + h;
    }

    private double pointHeight(double _x, double _z)
    {
      double h = 0;
      int x = (int)Math.Floor(_x);
      int z = (int)Math.Floor(_z);
      double tx = _x - x;
      double tz = _z - z;
      if (tx + tz > 1)
      { // top right triangle
        h = getHeight(1 - Math.Abs(_x - x), 1 - Math.Abs(_z - z), x, z + 1, x + 1, z + 1, x + 1, z);
      }
      else
      { // bottom left triangle
        h = getHeight(Math.Abs(_x - x), Math.Abs(_z - z), x + 1, z, x, z, x, z + 1);
      }
      return h;
    }

    private double toHeightMapX(double x)
    {
      x = x + scale;
      x = x / ((scale * 2) / landscapesize);
      return x;
    }

    private double toHeightMapZ(double z)
    {
      z = z + scale;
      z = z / ((scale * 2) / landscapesize);
      return z;
    }

    public double getHeight(double x, double z)
    {
      double px = toHeightMapX(x);
      double pz = toHeightMapZ(z);
      return pointHeight(px, pz);
    }


  }
}

File: Server/wwwroot/script/landscape.js
/*
This file is part of Armored Deathmatch by Hans Milling.

Armored Deathmatch is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

Armored Deathmatch is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Armored Deathmatch.  If not, see <http://www.gnu.org/licenses/>.
	
*/

var landscapescale = 20000;
var lsize = 256;
var texture1;

var skidmarkscanvas = document.createElement('canvas');
skidmarkscanvas.width = 1024;
skidmarkscanvas.height = 1024;
var ctx = skidmarkscanvas.getContext('2d');
/*
ctx.fillStyle = "rgba(200,0,0)";
ctx.fillRect(10, 10, 185, 170);

ctx.fillStyle = "rgba(0, 0, 200, 0.5)";
ctx.fillRect(40, 50, 755, 950);
*/
function addLandscape(ls) {
  landscape = ls;
  //landscape = generateLandscape(lsize + 1);

  texture1 = new THREE.Texture(skidmarkscanvas);
  texture1.needsUpdate = true;

  var texture2 = THREE.ImageUtils.loadTexture("graphics/landscape.jpg");
  texture2.repeat.x = 32;
  texture2.repeat.y = 32;
  texture2.wrapT = THREE.RepeatWrapping;
  texture2.wrapS = THREE.RepeatWrapping;
  texture2.needsUpdate = true;

  var material1 = new THREE.MeshLambertMaterial({ ambient: 0xffffff, transparent: true, map: texture1 });
  var material2 = new THREE.MeshLambertMaterial({ ambient: 0xffffff, map: texture2 });
/*  material1.polygonOffset = true;
  material1.polygonOffsetFactor = -1;*/
/*  material1.depthTest = false;
  material1.depthWrite = false;*/

  var materials = [material1, material2];

  //var material = new THREE.MeshLambertMaterial({ color: 0xffffff, map: texture });
  var plane = new THREE.PlaneGeometry(2 * landscapescale, 2 * landscapescale, lsize, lsize);

  for (var i = 0, l = plane.vertices.length; i < l; i++) {

    var x = i % (lsize + 1)
    var y = ~ ~(i / (lsize + 1));
    //landscape[x][y] = 0; //REMOVE THIS
    plane.vertices[i].z = landscape[x][y];
  }
  //plane.computeCentroids();

  var mesh = THREE.SceneUtils.createMultiMaterialObject(plane, materials);
  mesh.rotation.x = -90 * Math.PI / 180;
  scene.add(mesh);
}

function getLandscapeNormal(id) {
  var vidx = [11, 10, 5, 4]; // The four vertexes in the tankbody to use as height measure points 
  var p = [];
  for (var i = 0; i < 4; i++) {
    var r = tanks[id].rotation;
    var ox = tanks[id].tankbody.geometry.vertices[vidx[i]].z;
    var oz = tanks[id].tankbody.geometry.vertices[vidx[i]].x;
    var sx = tanks[id].tankbody.position.x + Math.sin(r) * ox + Math.cos(r) * oz;
    var sz = tanks[id].tankbody.position.z + Math.cos(r) * ox - Math.sin(r) * oz;
    var gx = toHeightMapX(sx);
    var gz = toHeightMapZ(sz);
    p[i] = new THREE.Vector3(sx, pointHeight(gx, gz), sz);
  }

  var N = new THREE.Vector3(
        (p[1].y - p[2].y) * (p[1].z - p[3].z) - (p[1].z - p[2].z) * (p[1].y - p[3].y),
        -((p[1].z - p[2].z) * (p[1].x - p[3].x) - (p[1].x - p[2].x) * (p[1].z - p[3].z)),
        (p[1].x - p[2].x) * (p[1].y - p[3].y) - (p[1].y - p[2].y) * (p[1].x - p[3].x)
      );
  N = N.normalize();
  N.x = -N.x;
  N.z = -N.z;
  return N;
}

function getHeight(px, pz, x1, z1, x2, z2, x3, z3) {
  var h1 = landscape[x1][z1];
  var h2 = landscape[x2][z2];
  var h3 = landscape[x3][z3];
  var diffx = h1 - h2;
  var diffz = h3 - h2;
  var h = px * diffx + pz * diffz;
  return h2 + h;
}

function pointHeight(_x, _z) {
  var h = 0;
  var x = Math.floor(_x);
  var z = Math.floor(_z);
  var tx = _x - x;
  var tz = _z - z;
  var d = document.getElementById("infowindow");
  if (tx + tz > 1) { // top right triangle
    h = getHeight(1 - Math.abs(_x - x), 1 - Math.abs(_z - z), x, z + 1, x + 1, z + 1, x + 1, z);
  } else { // bottom left triangle
    h = getHeight(Math.abs(_x - x), Math.abs(_z - z), x + 1, z, x, z, x, z + 1);
  }
  return h;
}

function toHeightMapX(x) {
  x = x + landscapescale;
  x = x / ((landscapescale * 2) / lsize);
  return x;
}

function toHeightMapZ(z) {
  z = z + landscapescale;
  z = z / ((landscapescale * 2) / lsize);
  return z;
}

function toSkidmarksX(x) {
  x += landscapescale;
  x = x / (landscapescale/512.0);
  return x;
}

function toSkidmarksZ(z) {
  z += landscapescale;
  z = z / (landscapescale/512.0);
  return z;
}
